src/Microsoft.Identity.Client/Azure/ServicePrincipal.cs

```diff
@@ -0,0 +1,248 @@
+using System.Globalization;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client.Azure
+{
+#if !ANDROID_BUILDTIME && !iOS_BUILDTIME && !WINDOWS_APP_BUILDTIME && !MAC_BUILDTIME // Hide confidential client on mobile platforms
+
+    /// <summary>
+    ///     ServicePrincipalProbe looks to the application setting and environment variables to build a ICredentialProvider.
+    /// </summary>
+    public class ServicePrincipalProbe : IProbe
+    {
+        private readonly IServicePrincipalConfiguration _config;
+
+        /// <summary>
+        /// Create a new instance of a ServicePrincipalProbe
+        /// </summary>
+        /// <param name="config">optional configuration; if not specified the default configuration will use environment variables</param>
+        public ServicePrincipalProbe(IServicePrincipalConfiguration config = null)
+        {
+            _config = config ?? new DefaultServicePrincipalConfiguration();
+        }
+
+        // Async method lacks 'await' operators and will run synchronously
+        /// <summary>
+        /// Check if the probe is available for use in the current environment
+        /// </summary>
+        /// <returns>True if a credential provider can be built</returns>
+        public Task<bool> AvailableAsync() => Task.FromResult(IsClientSecret() || IsClientCertificate());
+
+        /// <summary>
+        /// Create a credential provider from the information discovered by the probe
+        /// </summary>
+        /// <returns>A service principal credential provider</returns>
+        public async Task<ITokenProvider> ProviderAsync()
+        {
+            var available = await AvailableAsync().ConfigureAwait(false);
+            if (!available)
+                throw new InvalidOperationException("The required environment variables are not available.");
+
+            var authorityWithTenant = string.Format(CultureInfo.InvariantCulture, AadAuthority.AADCanonicalAuthorityTemplate, _config.Authority, _config.TenantId);
+
+            if (!IsClientCertificate())
+            {
+                return new ServicePrincipalTokenProvider(authorityWithTenant, _config.TenantId, _config.ClientId, _config.ClientSecret);
+            }
+
+            ClientAssertionCertificate cert;
+            if (!string.IsNullOrWhiteSpace(_config.CertificateBase64))
+            {
+                // If the certificate is provided as base64 encoded string in env, decode and hydrate a x509 cert
+                var decoded = Convert.FromBase64String(_config.CertificateBase64);
+                cert = new ClientAssertionCertificate(new X509Certificate2(decoded));
+            }
+            else
+            {
+                // Try to use the certificate store
+                var store = new X509Store(StoreNameWithDefault, StoreLocationFromEnv);
+                store.Open(OpenFlags.ReadOnly);
```

Comment:
- Don't we first rule out the expired certificates? 
- would we also want to find a certificate by `SubjectDistinguishedName`